scripts/fuzz_opt.py

```diff
@@ -145,83 +143,135 @@ def run_bynterp(wasm):
 
 
-def run_wasm2js(wasm):
-  wrapper = run([in_bin('wasm-opt'), wasm, '--emit-js-wrapper=/dev/stdout'] + FEATURE_OPTS)
-  cmd = [in_bin('wasm2js'), wasm, '--emscripten']
-  if random.random() < 0.5:
-    cmd += ['-O']
-  main = run(cmd + FEATURE_OPTS)
-  with open(os.path.join(options.binaryen_root, 'scripts', 'wasm2js.js')) as f:
-    glue = f.read()
-  with open('js.js', 'w') as f:
-    f.write(glue)
-    f.write(main)
-    f.write(wrapper)
-  out = fix_output(run_vm([NODEJS, 'js.js', 'a.wasm']))
-  if 'exception' in out:
-    # exception, so ignoring - wasm2js does not have normal wasm trapping, so opts can eliminate a trap
-    out = IGNORE
-  return out
-
-
-def run_vms(prefix):
-  wasm = prefix + 'wasm'
-  results = []
-  results.append(run_bynterp(wasm))
-  results.append(fix_output(run_vm([os.path.expanduser('d8'), prefix + 'js'] + V8_OPTS + ['--', wasm])))
-  if WASM2JS:
-    results.append(run_wasm2js(wasm))
-
-  # append to add results from VMs
-  # results += [fix_output(run_vm([os.path.expanduser('d8'), prefix + 'js'] + V8_OPTS + ['--', prefix + 'wasm']))]
-  # results += [fix_output(run_vm([os.path.expanduser('~/.jsvu/jsc'), prefix + 'js', '--', prefix + 'wasm']))]
-  # spec has no mechanism to not halt on a trap. so we just check until the first trap, basically
-  # run(['../spec/interpreter/wasm', prefix + 'wasm'])
-  # results += [fix_spec_output(run_unchecked(['../spec/interpreter/wasm', prefix + 'wasm', '-e', open(prefix + 'wat').read()]))]
-
-  if len(results) == 0:
-    results = [0]
-
-  # NaNs are a source of nondeterminism between VMs; don't compare them
-  if not NANS:
-    first = results[0]
-    for i in range(len(results)):
-      compare(first, results[i], 'comparing between vms at ' + str(i))
-
-  return results
+# Each test case handler receives two wasm files, one before and one after some changes
+# that should have kept it equivalent. It also receives the optimizations that the
+# fuzzer chose to run.
+class TestCaseHandler:
+  # If the core handle_pair() method is not overridden, it calls handle_single()
+  # on each of the pair. That is useful if you just want the two wasms, and don't
+  # care about their relationship
+  def handle_pair(self, before_wasm, after_wasm, opts):
+    self.handle(before_wasm, opts)
+    self.handle(after_wasm, opts)
+
+
+# Run VMs and compare results
+class CompareVMs(TestCaseHandler):
+  def handle_pair(self, before_wasm, after_wasm, opts):
+    run([in_bin('wasm-opt'), before_wasm, '--emit-js-wrapper=a.js', '--emit-spec-wrapper=a.wat'])
+    run([in_bin('wasm-opt'), after_wasm, '--emit-js-wrapper=b.js', '--emit-spec-wrapper=b.wat'])
+    before = self.run_vms('a.js', before_wasm)
+    after = self.run_vms('b.js', after_wasm)
+    self.compare_vs(before, after)
+
+  def run_vms(self, js, wasm):
+    results = []
+    results.append(run_bynterp(wasm))
+    results.append(fix_output(run_vm([os.path.expanduser('d8'), js] + V8_OPTS + ['--', wasm])))
```

Comment:
os.path.expanduser isn't doing anything here.

Reply:
I'm not sure how this works actually since this ends up calling subprocess.check_output, which I think expects an absolute path for argv0.

Reply:
It does work if `d8` is in the path, I tested.